### PR TITLE
local rate limit: follow up to #21327

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Change to unsigned, added wrap-around, more comments.
Additional Description:
Risk Level: low
Testing: regression testing
Docs Changes: no
Release Notes: no
Platform Specific Features: no

### CODE REVIEW

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc [hunk 1]
```diff
@@ -115,11 +115,17 @@ void LocalRateLimiterImpl::onFillTimerHelper(TokenState& tokens,
...
   for (const auto& descriptor : descriptors_) {
-    if (counter_ % descriptor.multiplier_ == 0) {
+    if (refill_counter_ % descriptor.multiplier_ == 0) {
       onFillTimerHelper(*descriptor.token_state_, descriptor.token_bucket_);
+    } else {
+      all_refilled = false;
     }
   }
+  if (all_refilled) {
+    refill_counter_ = 0;
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:127]: It's more precise this way. Example:
timer_refill: 50ms
descriptor1 timer_refill: 100ms
descriptor2 timer_refill: 150ms

With this change, counter is reset at value 6 (least common multiple of 100/50 and 150/50). Descriptor2 timer is precise.

Prior, the counter is reset at value MAX_UINT64 + 1. That's not divisible by 6 so descriptor2 timer effectively hits sooner than it should.

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h [hunk 2]
```diff
@@ -36,11 +36,16 @@ class LocalRateLimiterImpl {
     mutable std::atomic<uint32_t> tokens_;
     MonotonicTime fill_time_;
   };
-  int64_t counter_{0};
+  // Refill counter is incremented per each refill timer hit. Descriptors are
+  // refilled whenever the counter is even modulo the fraction of the
+  // descriptor refill interval over the timer refill interval. The counter is
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h:41]: Will do.

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:127]: OK maybe just add more comments/example? Sorry this is pretty confusing to me.

[the remark this answers is not in the document]
Comment (wbpcode) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:127]: I am also confused by this. Is this reset necessary? `MAX_UINT64` is a value we almost can never reached.

[the remark this answers is not in the document]
Comment (kyessenov) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc]: Removed this logic in the interest of readability. It's not that important.

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h [hunk 3]
```diff
@@ -36,11 +36,16 @@ class LocalRateLimiterImpl {
...
+  // reset if all descriptors are refilled.
+  uint64_t refill_counter_{0};
   struct LocalDescriptorImpl : public RateLimit::LocalDescriptor {
     std::shared_ptr<TokenState> token_state_;
     RateLimit::TokenBucket token_bucket_;
-    int64_t multiplier_;
+    // Descriptor refill interval is a multiple of the timer refill interval.
+    uint64_t multiplier_;
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h:48]: Added comment in both places.

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc [hunk 4]
```diff
@@ -48,6 +48,7 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
       throw EnvoyException(
           "local rate descriptor limit is not a multiple of token bucket fill timer");
     }
+    // Save the multiplicative factor to control the descriptor refill frequency.
```
Comment (mattklein123) [inline source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:51]: Sorry can you rename token_bucket to per_descriptor_token_bucket or something? I think that was the part that confused me the most. Then the rest makes sense, thank you.

/wait

File: source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc [hunk 5]
```diff
@@ -48,11 +48,12 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
...
     new_descriptor.multiplier_ = token_bucket.fill_interval_ / token_bucket_.fill_interval_;
```
Comment (mattklein123) [inline source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:52]: Sorry I meant the local in this function to make it more clear what this match is doing.

```suggestion
    new_descriptor.multiplier_ = per_descriptor_token_bucket.fill_interval_ / token_bucket_.fill_interval_;
```

/wait

**Comment by mattklein123** [comment]
LGTM but you need to fix DCO, sorry.

/wait

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**